### DISCUSSION TITLE
"Key not found, did not record episode"

### DISCUSSION THREAD

**Discussion by BEisem** [Taxel/PlexTraktSync#208] [Q&A] [answered]
In the logs, what does it mean when it says `Key not found, did not record episode`?

I'm getting this for several TV shows. I've checked both Plex library and Trakt, and the show (and episodes) exist in both locations. So what is causing this, and how do I fix?

**Comment by glensc** [comment]
@BEisem it's some weird code, it tries to lookup episode from trakt, but does not find it. as I recently refactored that part, it left me to scratch my head as well.

- current code: https://github.com/Taxel/PlexTraktSync/blob/e578129cb463455e6b68d5e138992fb9dfab0e62/plex_trakt_sync/commands/sync.py#L120
- it was added in this commit: 7a2c0a00afb8ebccd1100f0f12eef4c79036e652, unfortunately the commit message does not explain anything about the problem being solved and the commit does not seem to belong any issue or pull request

**Comment by simonc56** [comment]
Can you show full log please ?

**Reply by wandabastyle** [reply]
Like I explained in #172 this is partly related to episode ordering. I did more testing and with "TVDB sorting" I get matching guids for ep. 22, which I did not get before bcs. TMDB (there it is ep. 21) seems to have a different order for the series. Can't they for sure if this is right, bcs. I did the testing with a dummy file, so I don't know the content of the episode, just the title.

**Reply by glensc** [reply]
@wandabastyle there's no error in #172, are @wandabastyle and @BEisem the same person?

and please try to be concrete, link to the specific show on trakt and imdb/tvdb to be able to compare!

**Reply by BEisem** [reply]
No, we're not the same person. I can't speak for @wandabastyle but I'm very new at using github and I'm just doing my best. I'm not a coder, just someone who wants to help make this project better. So apologies, but I'm doing my best.

Here's an example of one show that is showing this problem.

https://www.imdb.com/title/tt10584350/
https://www.thetvdb.com/series/mapleworth-murders

There are entries in both IMDB and TVDB. But in the logs, I'm seeing this:

```2021-04-10 10:14:42,906 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E01
2021-04-10 10:14:42,906 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E02
2021-04-10 10:14:42,906 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E03
2021-04-10 10:14:42,906 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E04
2021-04-10 10:14:42,906 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E05
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E06
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E07
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E08
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E09
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E10
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E11
2021-04-10 10:14:42,907 WARNING:Show [Mapleworth Murders (2020)]: Key not found, did not record episode S01E12
```

**Reply by glensc** [reply]
I think the problem is simple:
- trakt search result is cached in sqlite

I've made a test that you could try running on your machine:
- https://github.com/Taxel/PlexTraktSync/pull/215

so, if the test fails, try removing `trakt_cache.sqlite` and re-run the test.

I'm sure this resolves your issue.

the proper fix would be to include TTL for cache entries, not use the default "forever"

**Reply by glensc** [reply]
@BEisem you have not responded here. Is it a cache issue here?

**Reply by BEisem** [reply]
I'm sorry, I've been unable to install pytest. I'm on a Mac, and even after trying all the installation methods available, pytest is not recognized. However I have tried deleting the `trakt_cache.sqlite` file and starting over, and the same episodes still result in the `Key not found` error.

**Reply by glensc** [reply]
unable to help with pytest, as you have not given details of the error, or how you attempted to install it and what's the install errors.

to my knowledge, it's bundled with python3. so would not need to install anything. but of course depends on distribution, linux distributions typically bundle development and runtime into different packages.

if `pytest` wrapper is not available, `python3 -m pytest` should behave identically. 

however, on my macOS with python3 from brew, seems pytest is installed via pip:

```
$ python3 -m pip list | grep test
pytest            6.2.2
```

**Comment by wandabastyle** [comment]
Yeah, same here. I just try to be helpful.
If this is related to the beta agent can you check and/or share the XML for one of the episodes?

@glensc Sorry I don't know where to start to comment, there are multiple discussions or pull requests which all can have the same root cause, the beta agent and missing or wrong metadata related. So it's kinda hard to bundle them.

**Reply by glensc** [reply]
@wandabastyle why don't you then read the whole discussion first and test the solutions? and report your outcome?
- https://github.com/Taxel/PlexTraktSync/discussions/208#discussioncomment-594359

if it's new problem, create a new issue, not hijack unrelated discussions. read the top post, this discussion is about "Key not found, did not record episode" error!

**Comment by flammable** [comment]
The issue might be that Trakt is moving from thetvdb to TMDB:

https://blog.trakt.tv/tmdb-transition-ef3d19a5cf24

I'm seeing the same thing, sometimes.

**Reply by glensc** [reply]
tl;dr, can you summarize what is the current default?

**Reply by BEisem** [reply]
> tl;dr, can you summarize what is the current default?

The basic situation is, they used to use TVDB but they are slowly transitioning to TMDB because TVDB is going to a paid subscription model.

From the article:

```
What’s the plan?
New TV shows are imported using TMDB as their data source (all new shows since November 2020 already use TMDB)
When we do nightly data refreshes, we compare TMDB with TVDB to determine if the seasons and episodes match up. (we’ve gathered data on over 20,000 shows since November 2020)
We created a comparison page for TMDB and TVDB data. This helps determine what data is missing or needs to be updated at TMDB.
If shows 100% match up, we will automatically change to TMDB as the data source. We’re also working on tools to help Trakt users verify and indicate a show can be switched to TMDB. (this will happen soon)
```

**Reply by BEisem** [reply]
And here's a link to the comparison page:

https://trakt.tv/transition/tmdb/updates

**Comment by glensc** [comment]
probably need to look on things case by case.

**Reply by glensc** [reply]
what I found that some episodes are just missing in trakt, for example, beavis and butt-head season 8:
- https://trakt.tv/shows/beavis-and-butt-head/seasons/7

![image](https://user-images.githubusercontent.com/199095/114754539-5b0e8980-9d61-11eb-95d9-a05cd0f69bfa.png)

**Reply by glensc** [reply]
the mappleworth murders that beisem reported, does seem to be his local problem, as it works for me locally:
- https://github.com/Taxel/PlexTraktSync/discussions/208#discussioncomment-594359

**Comment by BEisem** [comment]
FWIW, this error has now changed with the new version... now it says "Not found on Trakt." Which is a better, more descriptive error.

I've also noticed that this has brought to light some mismatch problems in my library, which is helpful. But it also brings to light certain problems where TVDB/TMDB/IMDB do not match or agree with each other. This is because Sonarr uses TVDB when creating the content, I believe Radarr uses TMDB, and Plex uses IMDB. So by the time the content gets to Plex, it might not be labeled or matched correctly.

One TV show I'm having a particular problem with is "Whose Line Is It Anyway." Every single episode is giving an error when being scanned by plextraktssync.

```
2021-04-14 18:35:52,319 WARNING[PlexTraktSync]:Skipping <imdb:tt14116976:<Episode:224716:Whose-Line-Is-It-Any-s17e06>>: Not found on Trakt
2021-04-14 18:35:52,359 WARNING[PlexTraktSync]:Skipping <imdb:tt14136386:<Episode:224717:Whose-Line-Is-It-Any-s17e07>>: Not found on Trakt
2021-04-14 18:35:52,399 WARNING[PlexTraktSync]:Skipping <tvdb:8202026:<Episode:224718:Whose-Line-Is-It-Any-s17e08>>: Not found on Trakt
2021-04-14 18:35:52,447 WARNING[PlexTraktSync]:Skipping <tvdb:8244885:<Episode:228073:Whose-Line-Is-It-Any-s17e09>>: Not found on Trakt
```

You can see that there are some that list IMDB, and some that list TVDB.

This show aired from 1998 to 2005, then there was a new version that started in 2013 and is still airing. IMDB and TMDB agree and call it one show. But TVDB splits it and considers it two different shows. Because of this, it's imported into Plex as a single show, but plextraktsync doesn't recognize it, and all the episodes throw the error.

I guess the moral of this story is, if there's any way to have the scanner look at *all* possible sources and try to match, maybe it would have better luck and more of a chance of matching. I don't even know if this is something that can be fixed, since the different sources don't match.

**Reply by glensc** [reply]
I have a plan to match all plex providers. plex has multiple providers under `.guid` attribute. currently plex-trakt-sync takes first, which is typically imdb or tvdb.

may need help designing logic how to matching using multiple providers. 

possible scenarios:
1. take first, iterate over `.guids` and take first that matches
2. iterate over `.guids`, lookup match for all of them.

for first, how reliable is it? is there provided that usually works? prefer that provider?
for second, what if t he outcome points to different trakt id? how do you resolve that? :)

**Reply by glensc** [reply]
I added `inspect` command, which is useful to see items `.guid` and `.guids` values:
- https://github.com/Taxel/PlexTraktSync/pull/237

**Reply by BEisem** [reply]
Nice, this is definitely helpful.

I tried it out with a random episode from Whose Line Is It Anyway that was throwing the error during scan.

```
% python3 main.py inspect 107583
Guid: 'plex://episode/5f6eeacef667c50041dd192f'
Guids: [<Guid:imdb://tt13158380>, <Guid:tvdb://7957625>]
Audio: 'stereo', 'English (AAC Stereo)'
Video: 'h264'
```

On IMDB, this correctly shows the episode: https://www.imdb.com/title/tt13158380/
But on TVDB, this is really odd... when I search for that GUID/TVDB number, it shows a person, not an episode. https://www.thetvdb.com/search?query=7957625

**Reply by glensc** [reply]
You need to search id of a certain type, as the id's overlap, i.e there can be a person and show with the same id.

However, I don't know how to do such a search.

**Reply by glensc** [reply]
Added trakt match into inspect command output.

Currently, it searches from the first guid, so it should match by imdb:
- https://trakt.tv/search/imdb?query=tt13158380

**Reply by BEisem** [reply]
Ok, this is great ... but it makes the mystery even greater as to why it can't find a match in Trakt.

Here's an example with one of the episodes of Whose Line Is It Anyway. This is the error in plextraktsync:

`2021-04-14 16:04:53,631 WARNING[PlexTraktSync]:Skipping <imdb:tt0747998:<Episode:107923:Whose-Line-Is-It-Any-s01e01>>: Not found on Trakt`

With this information, I use the new command:

```
% python3 main.py inspect 107923
Guid: 'plex://episode/5d9c103b2192ba001f367deb'
Guids: [<Guid:imdb://tt0747998>, <Guid:tmdb://995405>, <Guid:tvdb://216169>]
Audio: 'stereo', 'Unknown (MP3 Stereo)'
Video: 'mpeg4'
Provider: imdb
Id: tt0747998
Trakt match: <TVEpisode>: {'title': 'Whose Line is it Anyway?', 'year': 1998, 'ids': {'trakt': 60544, 'slug': 'whose-line-is-it-anyway-1998', 'tvdb': 76808, 'imdb': 'tt2919910', 'tmdb': 61018, 'tvrage': 6597}} S1E1 Greg Proops
```

So you can see that it does find a Trakt ID, which is 60544, and that matches the correct show. So I'm confused as to why the scan returns errors for each one of these episodes.

**Reply by BEisem** [reply]
Here's another one, just to show that it's not limited to just that show.

`2021-04-14 16:12:22,796 WARNING[PlexTraktSync]:Skipping <tmdb:2778937:<Episode:27214:Grace-s01e01>>: Not found on Trakt`

```
% python3 main.py inspect 27214
Guid: 'plex://episode/606eb2c447840e00356ff3ad'
Guids: [<Guid:tmdb://2778937>, <Guid:tvdb://8252676>]
Audio: 'stereo', 'English (EAC3 Stereo)'
Video: 'h264'
Provider: tmdb
Id: 2778937
Trakt match: <TVEpisode>: {'title': 'Grace', 'year': 2021, 'ids': {'trakt': 174611, 'slug': 'grace-2021', 'tvdb': 397762, 'imdb': 'tt13880508', 'tmdb': 120317, 'tvrage': None}} S1E1 Dead Simple
```

So when I do the `inspect` command, it finds all the information. But during the scan, it reports `Not found on Trakt.`

**Reply by glensc** [reply]
The difference can be explained, as the sync code uses different method:
- https://github.com/Taxel/PlexTraktSync/blob/22a1ef60cab82a7bf02d6b04b7d2cb0551b894f4/plex_trakt_sync/pytrakt_extensions.py#L16-L27

I can not answer at this time why that API call gives different results. from the brief look, it seems to find episodes by show.id, while the inspect command uses episode.id

**Reply by BEisem** [reply]
If I understand this correctly, it seems like maybe your method for the inspect command seems to be more robust. Can the sync command use episode.id instead of show.id?

**Reply by glensc** [reply]
I can not answer at this time. I don't know what part of the code.

however, I asked for those methods to be supported by pytrakt library, let's see where that goes:
- https://github.com/moogar0880/PyTrakt/issues/145

**Reply by glensc** [reply]
> But on TVDB, this is really odd... when I search for that GUID/TVDB number, it shows a person, not an episode. [thetvdb.com/search?query=7957625](https://www.thetvdb.com/search?query=7957625)

search by tvdb id:
- https://trakt.tv/search/tvdb?query=7957625

**Reply by glensc** [reply]
I guess the lookup table API method was created as the legacy agent had the id in form of show/season/episode format:

```
➔ ./plex_trakt_sync.sh inspect 7196
Guid: 'com.plexapp.agents.thetvdb://77137/1/1?lang=en'
Guids: []
Audio: 'stereo', 'Unknown (AC3 Stereo)'
Video: 'mpeg4'
Provider: tvdb
Id: 77137/1/1
Error: Method not Allowed
```

- https://trakt.tv/search/tvdb?query=77137

**Comment by glensc** [comment]
Re: https://github.com/Taxel/PlexTraktSync/discussions/208#discussioncomment-611321

> ```
> 2021-04-14 18:35:52,319 WARNING[PlexTraktSync]:Skipping <imdb:tt14116976:<Episode:224716:Whose-Line-Is-It-Any-s17e06>>: Not found on Trakt
> 2021-04-14 18:35:52,359 WARNING[PlexTraktSync]:Skipping <imdb:tt14136386:<Episode:224717:Whose-Line-Is-It-Any-s17e07>>: Not found on Trakt
> 2021-04-14 18:35:52,399 WARNING[PlexTraktSync]:Skipping <tvdb:8202026:<Episode:224718:Whose-Line-Is-It-Any-s17e08>>: Not found on Trakt
> 2021-04-14 18:35:52,447 WARNING[PlexTraktSync]:Skipping <tvdb:8244885:<Episode:228073:Whose-Line-Is-It-Any-s17e09>>: Not found on Trakt
> ```
> 
> You can see that there are some that list IMDB, and some that list TVDB.

The error may go away, if you take that show and do "Fix Match", then all episodes are of the same provider.

but don't do that yet, let's see if we can fix the code first.

**Comment by glensc** [comment]
Added PR that can use `.guid` from Plex Episode object, and is able to extract show in form of show/season/episode:
- https://github.com/Taxel/PlexTraktSync/pull/240

**Reply by BEisem** [reply]
Is this intended to be used with a sync? I downloaded to a new directory, copied `.env`, `.pytrakt.json` and `config.json` from a working version, and when I run `python3 main.py sync --sync=tv` I get the following crash:

```
bink@Homey-McPlexFace plex-trakt-sync-ep-lookup-quick % python3 main.py sync --sync=tv
INFO: Syncing with Plex BlaEis and Trakt binkatl
INFO: Syncing TV=True, Movies=False
INFO: Loaded Trakt lists in 3.8 seconds
INFO: Server version 1.22.2.4282-a97b03fad updated at: 2021-04-15 09:18:56
INFO: Recently added: [<Movie:123435:Billy-Connolly:-Live>, <Season:198033:Younger-s7>, <Season:198030:Chopped-s49>, <Movie:198006:Nobody>, <Movie:198004:At-First-Sight>]
Traceback (most recent call last):
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/trakt_api.py", line 208, in find_by_media
    search = trakt.sync.search_by_id(media_id, id_type=pm.provider, media_type=media_type)
  File "/Library/Python/3.7/site-packages/trakt/core.py", line 542, in inner
    resp = self._get_first(f, *args, **kwargs)
  File "/Library/Python/3.7/site-packages/trakt/core.py", line 489, in _get_first
    uri = next(generator)
  File "/Library/Python/3.7/site-packages/trakt/sync.py", line 255, in search_by_id
    raise ValueError('search_type must be one of {}'.format(valids))
ValueError: search_type must be one of ('trakt', 'trakt-movie', 'trakt-show', 'trakt-episode', 'trakt-person', 'imdb', 'tmdb', 'tvdb')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "main.py", line 6, in <module>
    cli()
  File "/Library/Python/3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Library/Python/3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Library/Python/3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Library/Python/3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Library/Python/3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/commands/sync.py", line 202, in sync
    sync_all(movies=movies, tv=tv)
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/commands/sync.py", line 167, in sync_all
    for tm, pe, te in for_each_episode(plex.show_sections, trakt):
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/commands/sync.py", line 127, in for_each_episode
    te = trakt.find_episode(pe, pm)
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/trakt_api.py", line 226, in find_episode
    return self.find_by_media(pe)
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/decorators/rate_limit.py", line 43, in wrapper
    return fn(*args, **kwargs)
  File "/Users/bink/scripts/plex-trakt-sync-ep-lookup-quick/plex_trakt_sync/trakt_api.py", line 213, in find_by_media
    raise ValueError(f"Invalid id_type: '{pm.provider}', guid: '{pm.guid}', guids: '{pm.guids}': {e}") from e
ValueError: Invalid id_type: 'local', guid: 'local://113357', guids: '[]': search_type must be one of ('trakt', 'trakt-movie', 'trakt-show', 'trakt-episode', 'trakt-person', 'imdb', 'tmdb', 'tvdb')
```

**Reply by glensc** [reply]
> `ValueError: Invalid id_type: 'local', guid: 'local://113357', guids: '[]': search_type must be one of ('trakt', 'trakt-movie', 'trakt-show', 'trakt-episode', 'trakt-person', 'imdb', 'tmdb', 'tvdb')`

guid local is and can not be supported. fix your library.

I'll check meanwhile why those items were not suppressed.

My guess is that your Show is okay, but not individual episodes underneath. See similar issue:
- https://github.com/Taxel/PlexTraktSync/discussions/208#discussioncomment-613667

**Reply by glensc** [reply]
While at it, run `inspect 113357` on it.

I'll perhaps add more code to inspect to see if inspected item is episode, try to find its season and show and show their details as well. i.e to show that perhaps your episodes do not have same guid. for example my library some episodes were from arbigary show, as i moved things to subfolder Comedy and trakt put some items under show Comedy

**Reply by BEisem** [reply]
Yep, that's a local ... how can I determine what show it is so I can go in and fix it?

```
bink@Homey-McPlexFace plex-trakt-sync-ep-lookup-quick % python3 main.py inspect 113357
Guid: 'local://113357'
Guids: []
Audio: '5.1(side)', 'English (AC3 5.1)'
Video: 'h264'
Provider: local
Id: 113357
Error: Invalid id_type: 'local', guid: 'local://113357', guids: '[]': search_type must be one of ('trakt', 'trakt-movie', 'trakt-show', 'trakt-episode', 'trakt-person', 'imdb', 'tmdb', 'tvdb')
```

**Reply by glensc** [reply]
- [ep-lookup-quick 86db886] Skip episodes of provider none

so, it should now print skipped.

**Reply by BEisem** [reply]
Something is still not right ... I'm still getting the same error. And I'm sure I have the new version, I went to `plex_trakt_sync/commands/sync.py` and checked for the new code in https://github.com/Taxel/PlexTraktSync/commit/86db8867c664309acb387be27db92ef2981dc309 to make sure it is there...

**Reply by glensc** [reply]
86db886 was a quick fix, did not test it. may need to look closer to your error trace.

**Reply by BEisem** [reply]
No worries, just wanted to point out that it is still causing an error.

Is there a way I can find out what show/episode is local://113357 so I can fix it?

**Reply by glensc** [reply]
Ah, yes, it was a silly typo; [ep-lookup-quick 25d636c] fixup! Skip episodes of provider none

you can open plex web with the id:
1. open any item in your plex web that is local to your library
2. find from url `key=%2Flibrary%2Fmetadata%2F10285` replace `10285` with your id `113357`

I'm planning add more support to the inspect command, so maybe wait with actual fixing:
- https://github.com/Taxel/PlexTraktSync/discussions/208#discussioncomment-615768

**Reply by glensc** [reply]
this might help for now:
- [ep-lookup-quick c8529e0] Inspect: Add detail of plex media item

**Reply by glensc** [reply]
This for inspect command:
- [ep-lookup-quick 434c13c] Inspect: Fetch show/season if episode given

**Comment by BEisem** [comment]


**Reply by glensc** [reply]
https://github.com/Taxel/PlexTraktSync/pull/240#issuecomment-821129098 gives pointer where feature was added

**Comment by glensc** [comment, marked as answer]
This is resolved in latest 0.8.x version.
